pylot/prediction/r2p2_predictor_operator.py

```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
+        runtime = (time.time() - start_time) * 1000
+        self._csv_logger.debug("{},{},{},{:.4f}".format(time_epoch_ms(),
+                                                        timestamp.coordinates[0],
+                                                        'r2p2-runtime',
+                                                        runtime))
         prediction_stream.send(
             PredictionMessage(timestamp, obstacle_predictions_list))
 
-    def get_occupancy_grid(self, point_cloud):
+
+    def pad_trajectory(self, trajectory):
+        # Take the appropriate number of past steps as specified by flags. 
+        # If we have not seen enough past locations of the vehicle, pad the
+        # trajectory with the appropriate number of copies of the original
+        # locations.
+        num_past_locations = trajectory.shape[0]
+        if num_past_locations < self._flags.prediction_num_past_steps:
+            initial_copies = np.repeat([np.array(trajectory[0])],
+                                       self._flags.prediction_num_past_steps - num_past_locations,
+                                       axis=0)
+            trajectory = np.vstack(
+                (initial_copies, trajectory))
+        elif num_past_locations > self._flags.prediction_num_past_steps:
+            trajectory = trajectory[-self._flags.prediction_num_past_steps:]
+        assert trajectory.shape[0] == self._flags.prediction_num_past_steps
+        return trajectory
+
+
+
+    def _get_closest_vehicles(self, all_vehicles):
+        distances = [v.trajectory[-1].get_angle_and_magnitude(Location())[1]
+                        for v in all_vehicles]
+        # The first vehicle will always be the ego vehicle.
+        sorted_vehicles = [v for v, d in sorted(
+            zip(all_vehicles, distances), key=lambda pair: pair[1])
+            if d <= self._flags.prediction_radius]
+
+        if self._flags.prediction_ego_agent:
+            return sorted_vehicles, sorted_vehicles[0]
+        else: # Exclude the ego vehicle.
+            return sorted_vehicles[1:], sorted_vehicles[0]
+
+
+    def _get_closest_vehicles_ego_transforms(self, closest_vehicles, ego_vehicle):
+        closest_vehicles_ego_locations = np.stack([v.trajectory[-1] for v in closest_vehicles])
+        closest_vehicles_ego_transforms = []
+
+        # Add appropriate rotations to closest_vehicles_ego_transforms, which
+        # we estimate using the direction determined by the last two distinct
+        # locations
+        for i in range(len(closest_vehicles)):
+            cur_vehicle_angle = self._get_vehicle_orientation(closest_vehicles[i])
+            closest_vehicles_ego_transforms.append(
+                Transform(location=closest_vehicles_ego_locations[i].location, 
+                          rotation=Rotation(yaw=cur_vehicle_angle)
+            ))
+        return closest_vehicles_ego_transforms
+
+
+    def _get_vehicle_orientation(self, vehicle):
```

Comment:
Shouldn't this be a method of DetectedObstacle class?

Reply:
`_get_vehicle_orientation` uses the past locations of the vehicle to estimate its heading, so I'm not sure it fits in the DetectedObstacle class.